### DISCUSSION TITLE
`.debug()` has been renamed to `._printChanges()`

### DISCUSSION THREAD

**Discussion by oliverfoggin** [pointfreeco/swift-composable-architecture#1465] [Beta]
I just updated from the beta branch to the release version of TCA and it told me that `.debug()` has been deprecated and to use `_printChanges`.

Is that correct? It seems odd to have to use an underscore function.

When trying to use it it seems to cause the compiler to think it's an ambiguous statement.

Thanks

**Comment by pyrtsa** [comment]
I think there's a little design error in the new `_printChanges(_:)` reducer modifier in that it expects an `Optional<Printer>` of some `Printer: _ReducerPrinter` rather than just non-optional. And that trips up the static member lookup ([SE-0299](https://github.com/apple/swift-evolution/blob/main/proposals/0299-extend-generic-static-member-lookup.md#detailed-design)) which expects the types to match.

As a workaround, I found that you can call it like any of:
- `_printChanges(Optional(.actionLabels))`, or
- `_printChanges(.some(.actionLabels))`, or
- `_printChanges(_ActionLabelsPrinter.actionLabels)`

I think the library should reconsider that optional though, and simplify the signature down to:
```diff
   @inlinable
   public func _printChanges<Printer: _ReducerPrinter>(
-    _ printer: Printer?
+    _ printer: Printer
-  ) -> ReducerBuilder<State, Action>._Conditional<_PrintChangesReducer<Self, Printer>, Self>
+  ) -> _PrintChangesReducer<Self, Printer>
```

**Reply by oliverfoggin** [reply]
Ah! Excellent, thanks very much 👍🏻

**Reply by stephencelis** [reply]
> I think there's a little design error in the new `_printChanges(_:)` reducer modifier in that it expects an `Optional<Printer>` of some `Printer: _ReducerPrinter` rather than just non-optional. And that trips up the static member lookup ([SE-0299](https://github.com/apple/swift-evolution/blob/main/proposals/0299-extend-generic-static-member-lookup.md#detailed-design)) which expects the types to match.

Dang, I could have sworn I had something working, but indeed it looks broken now.

The thought was being able to support dynamically disabling things in a fashion similar to SwiftUI:

```swift
._printChanges(self.shouldDebug ? .actionLabels : nil)
```

But maybe this isn't possible, or maybe we should have an explicit: `.none`:

```swift
._printChanges(self.shouldDebug ? .actionLabels : .none)
```

**Reply by pyrtsa** [reply]
Well this isn't a strong part of design in SwiftUI either because you can only use conditional selection between static member lookup expressions when both values have the same type which, more often than not, is not the case in SwiftUI either, consider:

```swift
Button("Press") {
}
.buttonStyle(wantItBordered ? .bordered : .automatic) // nope 🛑
```

The just created PR looks like an improvement, no type system trickery. 👌

(Edit: Haven't considered if this could nowadays change with implictly opened existentials, but at least SwiftUI hasn't adopted those yet.)
